### PR TITLE
[SPARK-35359][SQL] Insert data with char/varchar datatype will fail when data length exceed length limitation

### PR DESCRIPTION
**Pull request by fhygh**
### What changes were proposed in this pull request?
This PR is used to fix this bug:

```
set spark.sql.legacy.charVarcharAsString=true;
create table chartb01(a char(3));
insert into chartb01 select 'aaaaa';
```

here we expect the data of table chartb01 is 'aaa', but it runs failed.

### Why are the changes needed?
Improve backward compatibility

```
spark-sql>
         > create table tchar01(col char(2)) using parquet;
Time taken: 0.767 seconds
spark-sql>
         > insert into tchar01 select 'aaa';
ERROR | Executor task launch worker for task 0.0 in stage 0.0 (TID 0) | Aborting task | org.apache.spark.util.Utils.logError(Logging.scala:94)
java.lang.RuntimeException: Exceeds char/varchar type length limitation: 2
        at org.apache.spark.sql.catalyst.util.CharVarcharCodegenUtils.trimTrailingSpaces(CharVarcharCodegenUtils.java:31)
        at org.apache.spark.sql.catalyst.util.CharVarcharCodegenUtils.charTypeWriteSideCheck(CharVarcharCodegenUtils.java:44)
        at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.project_doConsume_0$(Unknown Source)
        at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
        at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
        at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:755)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$executeTask$1(FileFormatWriter.scala:279)
        at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1500)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.executeTask(FileFormatWriter.scala:288)
        at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$write$15(FileFormatWriter.scala:212)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
        at org.apache.spark.scheduler.Task.run(Task.scala:131)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1466)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

### Does this PR introduce _any_ user-facing change?
No (the legacy config is false by default).

### How was this patch tested?
Added unit tests.

### CODE REVIEW

**Comment by yaooqinn** [comment]
> We should update `TableOutputResolver.checkField` to not add the string length check expression if the legacy config is true.

+1

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 1]
```diff
@@ -805,6 +805,18 @@ class FileSourceCharVarcharTestSuite extends CharVarcharTestSuite with SharedSpa
     }
   }
 
+  test("create table and insert data over length values") {
```
[the remark this answers is not in the document]
Comment (fhygh) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:808]: ok,thank you.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java [hunk 2]
```diff
@@ -26,7 +27,7 @@ private static UTF8String trimTrailingSpaces(
       UTF8String inputStr, int numChars, int limit) {
     int numTailSpacesToTrim = numChars - limit;
     UTF8String trimmed = inputStr.trimTrailingSpaces(numTailSpacesToTrim);
-    if (trimmed.numChars() > limit) {
+    if (trimmed.numChars() > limit && !SQLConf.get().charVarcharAsString()) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:30]: We don't need this now.

Comment (fhygh) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:30]: then do we need update PartitioningUtils.charTypeWriteSideCheck and PartitioningUtils.varcharTypeWriteSideCheck ?

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:30]: How do we plan to update these two?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/PartitioningUtils.scala [hunk 3]
```diff
@@ -72,12 +73,20 @@ private[sql] object PartitioningUtils {
 
   private def charTypeWriteSideCheck(inputStr: String, limit: Int): String = {
     val toUtf8 = UTF8String.fromString(inputStr)
-    CharVarcharCodegenUtils.charTypeWriteSideCheck(toUtf8, limit).toString
+    if (SQLConf.get.charVarcharAsString) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/PartitioningUtils.scala:76]: can we move the check to `normalizePartitionSpec`?
```
val normalizedVal = if (SQLConf.get.charVarcharAsString) value else normalizedFiled.dataType match {
  ...
}
```

**Review by yaooqinn — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Pull request closed without merging**